File: Assignment_2_Florakis.py
###STATISTICAL MACHINE LEARNING ASSIGNMENT II

#Necessary Libraries
#!pip install sklearn
#!pip install matplotlib
#!pip install pandas
#!pip install seaborn
#!pip install tensorflow
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np 
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

# Load Data
data = pd.read_csv('diabetes.csv')

##Random Forest

#i) Random Forest for all variables

# Standardize the data
scale = StandardScaler()
columns_to_scale = data.columns[:-1]  # Exclude the 'Outcome' column
data[columns_to_scale] = scale.fit_transform(data[columns_to_scale])

# Split data into features (x) and target variable (y)
x = data.iloc[:, :-1]
y = data['Outcome']

# Split the scaled data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.8, random_state=12345)

# Initialize results list
results = []

# Define the number of trees
n_trees = [500, 1000, 2000]

# Loop through different numbers of trees
for n_tree in n_trees:
    # Create RandomForestClassifier with the specified number of trees
    rf_classifier = RandomForestClassifier(random_state=123, n_estimators=n_tree)
    
    # Perform cross-validation
    cv_scores = cross_val_score(rf_classifier, x_train, y_train, cv=5)
    
    # Fit the model on the training set
    rf_classifier.fit(x_train, y_train)
    
    # Predictions on the testing dataset
    predictions_train = rf_classifier.predict(x_train)
    
    # Calculate testing accuracy
    train_accuracy = accuracy_score(y_train, predictions_train)
    
    # Predictions on the testing dataset
    predictions_test = rf_classifier.predict(x_test)
    
    # Calculate testing accuracy
    test_accuracy = accuracy_score(y_test, predictions_test)
    
    # Append results to the list
    results.append([n_tree,cv_scores.mean(),test_accuracy])

    # Visualize feature importance
    feature_importances = pd.Series(rf_classifier.feature_importances_, index=x_train.columns).sort_values(ascending=False)
    plt.figure(figsize=(10, 6))
    sns.barplot(x=feature_importances, y=feature_importances.index)
    plt.xlabel('Feature Importance Score')
    plt.ylabel('Features')
    plt.title(f"Visualizing Important Features - {n_tree} Trees")
    plt.show()

# Convert the list to a DataFrame
results_df = pd.DataFrame(results, columns=['Number of Trees','Training Accuracy','Testing Accuracy'])

# Print the results
print(results_df)



#ii) Random Forest for “Glucose” & “BMI”

# Load Data
data = pd.read_csv('diabetes.csv')

# Select a subset of columns
subset_columns = ["Glucose", "BMI"]

# Standardize the subset data
scale = StandardScaler()
data[subset_columns] = scale.fit_transform(data[subset_columns])

# Use only the subset columns
x_subset = data[subset_columns]
y = data['Outcome']

# Split the scaled subset data into training and testing sets
x_train_subset, x_test_subset, y_train, y_test = train_test_split(x_subset, y, train_size=0.8, random_state=12345)

# Initialize results list
results_subset = []

# Define the number of trees
n_trees = [500, 1000, 2000]

# Loop through different numbers of trees
for n_tree in n_trees:
    # Create RandomForestClassifier with the specified number of trees
    rf_classifier_subset = RandomForestClassifier(random_state=123, n_estimators=n_tree)
    
    # Perform cross-validation on the subset
    cv_scores_subset = cross_val_score(rf_classifier_subset, x_train_subset, y_train, cv=5)
    
    # Fit the model on the training set (subset)
    rf_classifier_subset.fit(x_train_subset, y_train)
    
    # Predictions on the testing dataset (subset)
    predictions_train_subset = rf_classifier_subset.predict(x_train_subset)
    
    # Calculate testing accuracy (subset)
    train_accuracy_subset = accuracy_score(y_train, predictions_train_subset)
    
    # Predictions on the testing dataset (subset)
    predictions_test_subset = rf_classifier_subset.predict(x_test_subset)
    
    # Calculate testing accuracy (subset)
    test_accuracy_subset = accuracy_score(y_test, predictions_test_subset)
    
    # Append results to the list
    results_subset.append([n_tree,cv_scores_subset.mean(),test_accuracy_subset])

    # Visualize feature importance for the subset
    feature_importances_subset = pd.Series(rf_classifier_subset.feature_importances_, index=x_train_subset.columns).sort_values(ascending=False)
    plt.figure(figsize=(10, 6))
    sns.barplot(x=feature_importances_subset, y=feature_importances_subset.index)
    plt.xlabel('Feature Importance Score')
    plt.ylabel('Features')
    plt.title(f"Visualizing Important Features - {n_tree} Trees (Subset)")
    plt.show()

# Convert the list to a DataFrame for the subset
results_df_subset = pd.DataFrame(results_subset, columns=['Number of Trees', 'Training Accuracy', 'Testing Accuracy'])

# Print the results for the subset
print(results_df_subset)




##Feedforward Neural Network

# Load Data
data = pd.read_csv('diabetes.csv')

# Convert outcome to a numeric array and explicitly cast to int64
y = np.array(data['Outcome'], dtype=np.int64)

# Normalize data (variables only)
scale = StandardScaler()
X = scale.fit_transform(data.iloc[:, :8])

# Create training data as 80% of the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

# Build and train the neural network model
def build_and_train_model(neurons, learning_rate):
    model = Sequential([
        Dense(neurons, activation='relu', input_shape=(8,)),
        Dense(neurons, activation='relu'),
        Dense(1, activation='sigmoid')
    ])

    # Compile the model
    model.compile(
        loss='binary_crossentropy',
        optimizer=Adam(learning_rate=learning_rate),
        metrics=['accuracy']
    )

    # Train the model
    history = model.fit(
        x=X_train,
        y=y_train,
        epochs=300,
        batch_size=100,
        validation_split=0.2
    )

    # Evaluate the model on the test set
    test_metrics = model.evaluate(x=X_test, y=y_test)
    train_metrics = model.evaluate(x=X_train, y=y_train)

    # Plot training history
    plot_training_history(neurons, learning_rate, history)

    return neurons, learning_rate, train_metrics[1], test_metrics[1]

# Function to plot training history
def plot_training_history(neurons, learning_rate, history):
    plt.figure(figsize=(12, 6))

    # Plot training & validation accuracy values
    plt.subplot(1, 2, 1)
    plt.plot(history.history['accuracy'])
    plt.plot(history.history['val_accuracy'])
    plt.title(f'Model Accuracy - Neurons: {neurons}, LR: {learning_rate}')
    plt.xlabel('Epoch')
    plt.ylabel('Accuracy')
    plt.legend(['Train', 'Validation'], loc='upper left')

    # Plot training & validation loss values
    plt.subplot(1, 2, 2)
    plt.plot(history.history['loss'])
    plt.plot(history.history['val_loss'])
    plt.title(f'Model Loss - Neurons: {neurons}, LR: {learning_rate}')
    plt.xlabel('Epoch')
    plt.ylabel('Loss')
    plt.legend(['Train', 'Validation'], loc='upper left')

    plt.tight_layout()
    plt.show()

# Matrix for results
results = []

# Different parameters
neurons_list = [4, 8, 16]
learning_rates = [0.001, 0.01]

# Running the models and extracting results
for neurons in neurons_list:
    for lr in learning_rates:
        result = build_and_train_model(neurons, lr)
        results.append(result)

# Display a summary table of results
summary_table = pd.DataFrame(results, columns=['Neurons', 'Learning Rate', 'Train Accuracy', 'Test Accuracy'])
print(summary_table)





#ii) FNN for “Glucose” & “BMI”

# Load Data
data = pd.read_csv('diabetes.csv')

# Change types of variables to numeric and factor only
data['Outcome'] = pd.Categorical(data['Outcome']).codes

# Normalize data (variables only) and check the results
scale = StandardScaler()
data[['Glucose', 'BMI']] = scale.fit_transform(data[['Glucose', 'BMI']])

# Convert outcome to a numeric array and explicitly cast to int64
y = np.array(data['Outcome'], dtype=np.int64)

# Create a reproducible random sampling
np.random.seed(123)
random_sample = np.random.rand(len(data)) < 0.8
train_data = data[random_sample]
test_data = data[~random_sample]

# Build and train the neural network model
def build_and_train_model(neurons, learning_rate):
    model = Sequential()
    model.add(Dense(neurons, activation='relu', input_shape=(2,)))
    model.add(Dense(neurons, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))

    # Compile the model
    model.compile(
        loss='binary_crossentropy',
        optimizer='adam',
        metrics=['accuracy']
    )

    # Train the model
    history = model.fit(
        x=np.array(train_data[['Glucose', 'BMI']]),
        y=y[train_data.index],
        epochs=300,
        batch_size=100,
        validation_split=0.2
    )

    # Evaluate the model on the test set
    test_metrics = model.evaluate(
        x=np.array(test_data[['Glucose', 'BMI']]),
        y=y[test_data.index]
    )

    # Evaluate the model on the training set
    train_metrics = model.evaluate(
        x=np.array(train_data[['Glucose', 'BMI']]),
        y=y[train_data.index]
    )

    # Plot training history
    plot_training_history(neurons, learning_rate, history)

    return [neurons, learning_rate, train_metrics[1], test_metrics[1]]

# Function to plot training history
def plot_training_history(neurons, learning_rate, history):
    plt.figure(figsize=(12, 6))

    # Plot training & validation accuracy values
    plt.subplot(1, 2, 1)
    plt.plot(history.history['accuracy'])
    plt.plot(history.history['val_accuracy'])
    plt.title(f'Model Accuracy - Neurons: {neurons}, LR: {learning_rate}')
    plt.xlabel('Epoch')
    plt.ylabel('Accuracy')
    plt.legend(['Train', 'Validation'], loc='upper left')

    # Plot training & validation loss values
    plt.subplot(1, 2, 2)
    plt.plot(history.history['loss'])
    plt.plot(history.history['val_loss'])
    plt.title(f'Model Loss - Neurons: {neurons}, LR: {learning_rate}')
    plt.xlabel('Epoch')
    plt.ylabel('Loss')
    plt.legend(['Train', 'Validation'], loc='upper left')

    plt.tight_layout()
    plt.show()

# Matrix for results
results = np.empty((6, 4))
results[:] = np.nan
colnames = ["Neurons", "Learning Rate", "Train Accuracy", "Test Accuracy"]

# Different parameters
neurons_list = [4, 8, 16]
learning_rates = [0.001, 0.01]

# Running the models and extracting results
index = 0
for neurons in neurons_list:
    for lr in learning_rates:
        result = build_and_train_model(neurons, lr)
        results[index, :] = result
        index += 1

# Display a summary table of results
summary_table = pd.DataFrame(results, columns=colnames)
print(summary_table)

File: Assignment_2_Florakis.R
###STATISTICAL MACHINE LEARNING ASSIGNMENT II

#Necessary Libraries
library(caret)
library(randomForest)
library(keras)
library(tensorflow)
library(dplyr)

#import data
data<- read.csv(file = "diabetes.csv")
str(data)

#change types of variables ton numeric and factor only
data$Pregnancies <- as.numeric(data$Pregnancies)
data$Glucose <- as.numeric(data$Glucose)
data$BloodPressure <- as.numeric(data$BloodPressure)
data$SkinThickness <- as.numeric(data$SkinThickness)
data$Insulin <- as.numeric(data$Insulin)
data$Age <- as.numeric(data$Age)
data$Outcome = factor(data$Outcome, levels = c(0, 1))

#Normalize data (variables only) and check the results
data[-9]<-scale(data[-9])
str(data)

#View(data) #if need to visually check the whole dataset

##Random Forest

#i) Random Forest for all variables

# Reproducible random sampling
set.seed(123)

# Creating training data as 80% of the dataset
random_sample <- createDataPartition(data$Outcome, p = 0.80, list = F)

# Generating training dataset from the random_sample
training_dataset  <- data[random_sample, ]

# Generating testing dataset from rows which are not included in random_sample
testing_dataset <- data[-random_sample, ]

# NTree --> 500 
RF <- randomForest(Outcome ~ ., data = training_dataset, ntree = 500, cv.fold = 5)

# View the cross-validated forest results.
print(RF)

# Importance of each predictor.
importance(RF, type = 2)

# Plot the cross-validated random forest.
plot(RF)

# Variable importance plot for cross-validated random forest.
varImpPlot(RF)

# Make predictions on the training dataset.
predictions_train <- predict(RF, data = training_dataset)

# Calculate the accuracy.
accuracy_train <- sum(predictions_train == training_dataset$Outcome) / length(training_dataset$Outcome)

# Make predictions on the testing dataset.
predictions_test <- predict(RF, testing_dataset)

# Create a confusion matrix for the testing dataset.
confusion_matrix_test <- confusionMatrix(data = predictions_test, reference = testing_dataset$Outcome)

# Extract and print accuracy for the testing dataset.
accuracy_test <- confusion_matrix_test$overall[1]

# NTree --> 1000 
RF1 <- randomForest(Outcome ~ ., data = training_dataset, ntree = 1000, cv.fold = 5)

# View the cross-validated forest results.
print(RF1)

# Importance of each predictor.
importance(RF1, type = 2)

# Plot the cross-validated random forest.
plot(RF1)

# Variable importance plot for cross-validated random forest.
varImpPlot(RF1)

# Make predictions on the training dataset.
predictions_train1 <- predict(RF1, data = training_dataset)

# Calculate the accuracy.
accuracy_train1 <- sum(predictions_train1 == training_dataset$Outcome) / length(training_dataset$Outcome)

# Make predictions on the testing dataset.
predictions_test1 <- predict(RF1, testing_dataset)

# Create a confusion matrix for the testing dataset.
confusion_matrix_test1 <- confusionMatrix(data = predictions_test1, reference = testing_dataset$Outcome)

# Extract and print accuracy for the testing dataset.
accuracy_test1 <- confusion_matrix_test1$overall[1]

# NTree --> 2000 
RF2 <- randomForest(Outcome ~ ., data = training_dataset, ntree = 2000, cv.fold = 5)

# View the cross-validated forest results.
print(RF2)

# Importance of each predictor.
importance(RF2, type = 2)

# Plot the cross-validated random forest.
plot(RF2)

# Variable importance plot for cross-validated random forest.
varImpPlot(RF2)

# Make predictions on the training dataset.
predictions_train2 <- predict(RF2, data = training_dataset)

# Calculate the accuracy.
accuracy_train2 <- sum(predictions_train2 == training_dataset$Outcome) / length(training_dataset$Outcome)

# Make predictions on the testing dataset.
predictions_test2 <- predict(RF2, testing_dataset)

# Create a confusion matrix for the testing dataset.
confusion_matrix_test2 <- confusionMatrix(data = predictions_test2, reference = testing_dataset$Outcome)

# Extract and print accuracy for the testing dataset.
accuracy_test2 <- confusion_matrix_test2$overall[1]

#Creating a training and testing accuracy dataframe 
data.frame( modelName = c("RF   (NTree=500)","RF1 (NTree=1000)","RF2 (NTree=2000)"),
            AccuracyTrain=c(accuracy_train,accuracy_train1,accuracy_train2),
            AccuracyTest=c(accuracy_test,accuracy_test1,accuracy_test2))

#ii) Random Forest for “Glucose” & “BMI”

#Downsizing the dataset as requested
data<-subset(data,select = c(Glucose,BMI,Outcome))
str(data)

# Reproducible random sampling
set.seed(123)

# Creating training data as 80% of the dataset
random_sample <- createDataPartition(data$Outcome, p = 0.80, list = F)

# Generating training dataset from the random_sample
training_dataset  <- data[random_sample, ]

# Generating testing dataset from rows which are not included in random_sample
testing_dataset <- data[-random_sample, ]

# NTree --> 500 
RF <- randomForest(Outcome ~ ., data = training_dataset, ntree = 500, cv.fold = 5)

# View the cross-validated forest results.
print(RF)

# Importance of each predictor.
importance(RF, type = 2)

# Plot the cross-validated random forest.
plot(RF)

# Variable importance plot for cross-validated random forest.
varImpPlot(RF)

# Make predictions on the training dataset.
predictions_train <- predict(RF, data = training_dataset)

# Calculate the accuracy.
accuracy_train <- sum(predictions_train == training_dataset$Outcome) / length(training_dataset$Outcome)

# Make predictions on the testing dataset.
predictions_test <- predict(RF, testing_dataset)

# Create a confusion matrix for the testing dataset.
confusion_matrix_test <- confusionMatrix(data = predictions_test, reference = testing_dataset$Outcome)

# Extract and print accuracy for the testing dataset.
accuracy_test <- confusion_matrix_test$overall[1]

# NTree --> 1000 
RF1 <- randomForest(Outcome ~ ., data = training_dataset, ntree = 1000, cv.fold = 5)

# View the cross-validated forest results.
print(RF1)

# Importance of each predictor.
importance(RF1, type = 2)

# Plot the cross-validated random forest.
plot(RF1)

# Variable importance plot for cross-validated random forest.
varImpPlot(RF1)

# Make predictions on the training dataset.
predictions_train1 <- predict(RF1, data = training_dataset)

# Calculate the accuracy.
accuracy_train1 <- sum(predictions_train1 == training_dataset$Outcome) / length(training_dataset$Outcome)

# Make predictions on the testing dataset.
predictions_test1 <- predict(RF1, testing_dataset)

# Create a confusion matrix for the testing dataset.
confusion_matrix_test1 <- confusionMatrix(data = predictions_test1, reference = testing_dataset$Outcome)

# Extract and print accuracy for the testing dataset.
accuracy_test1 <- confusion_matrix_test1$overall[1]

# NTree --> 2000 
RF2 <- randomForest(Outcome ~ ., data = training_dataset, ntree = 2000, cv.fold = 5)

# View the cross-validated forest results.
print(RF2)

# Importance of each predictor.
importance(RF2, type = 2)

# Plot the cross-validated random forest.
plot(RF2)

# Variable importance plot for cross-validated random forest.
varImpPlot(RF2)

# Make predictions on the training dataset.
predictions_train2 <- predict(RF2, data = training_dataset)

# Calculate the accuracy.
accuracy_train2 <- sum(predictions_train2 == training_dataset$Outcome) / length(training_dataset$Outcome)

# Make predictions on the testing dataset.
predictions_test2 <- predict(RF2, testing_dataset)

# Create a confusion matrix for the testing dataset.
confusion_matrix_test2 <- confusionMatrix(data = predictions_test2, reference = testing_dataset$Outcome)

# Extract and print accuracy for the testing dataset.
accuracy_test2 <- confusion_matrix_test2$overall[1]

#Creating a training and testing accuracy dataframe 
data.frame( ModelName = c("RF   (NTree=500)","RF1 (NTree=1000)","RF2 (NTree=2000)"),
            AccuracyTrain=c(accuracy_train,accuracy_train1,accuracy_train2),
            AccuracyTest=c(accuracy_test,accuracy_test1,accuracy_test2))







##Feedforward Neural Network

#import data
data<- read.csv(file = "diabetes.csv")

#change types of variables ton numeric and factor only
data$Pregnancies <- as.numeric(data$Pregnancies)
data$Glucose <- as.numeric(data$Glucose)
data$BloodPressure <- as.numeric(data$BloodPressure)
data$SkinThickness <- as.numeric(data$SkinThickness)
data$Insulin <- as.numeric(data$Insulin)
data$Age <- as.numeric(data$Age)
data$Outcome <- as.factor(data$Outcome)

#Normalize data (variables only) and check the results
data[-9]<-scale(data[-9])
str(data)

#View(data) #if need to visually check the whole dataset

#i)FNN for all variables

# Create a reproducible random sampling
set.seed(123)

# Create training data as 80% of the dataset
random_sample <- createDataPartition(data$Outcome, p = 0.8, list = FALSE)
train_data <- data[random_sample, ]
test_data <- data[-random_sample, ]

# Build and train the neural network model
train_and_evaluate_model <- function(neurons, learning_rate) {
  model <- keras_model_sequential() %>%
    layer_dense(units = neurons, activation = "relu", input_shape = c(8)) %>%
    layer_dense(units = neurons, activation = "relu") %>%
    layer_dense(units = 1, activation = "sigmoid")
  
  # Compile the model
  model %>% compile(
    loss = "binary_crossentropy",
    optimizer = optimizer_adam(learning_rate = learning_rate),
    metrics = "accuracy"
  )
  
  # Train the model
  history <- model %>% fit(
    x = as.matrix(train_data[, 1:8]),
    y = as.numeric(train_data$Outcome) - 1,
    epochs = 300,
    batch_size = 100,
    validation_split = 0.2
  )
  
  # Evaluate the model on the test set
  test_metrics <- model %>% evaluate(
    x = as.matrix(test_data[, 1:8]),
    y = as.numeric(test_data$Outcome) - 1
  )
  
  # Evaluate the model on the training set
  train_metrics <- model %>% evaluate(
    x = as.matrix(train_data[, 1:8]),
    y = as.numeric(train_data$Outcome) - 1
  )
  
  return(c(neurons, learning_rate, train_metrics["accuracy"], test_metrics["accuracy"]))
}

# Matrix for results
results <- matrix(NA, nrow = 6, ncol = 4)
colnames(results) <- c("Neurons", "Learning Rate", "Train Accuracy", "Test Accuracy")

# Different parameters
neurons_list <- c(4, 8, 16)
learning_rates <- c(0.001, 0.01)

# Running the models and extracting results
index <- 1
for (neurons in neurons_list) {
  for (lr in learning_rates) {
    result <- train_and_evaluate_model(neurons, lr)
    results[index, ] <- result
    index <- index + 1
  }
}

# Display a summary table of results
summary_table <- as.data.frame(results)
print(summary_table)


#ii) FNN for “Glucose” & “BMI”

# Build and train the neural network model
train_and_evaluate_model <- function(neurons, learning_rate) {
  model <- keras_model_sequential() %>%
    layer_dense(units = neurons, activation = "relu", input_shape = c(2)) %>%
    layer_dense(units = neurons, activation = "relu") %>%
    layer_dense(units = 1, activation = "sigmoid")
  
  model %>% compile(
    loss = "binary_crossentropy",
    optimizer = optimizer_adam(learning_rate = learning_rate),
    metrics = "accuracy"
  )
  
  history <- model %>% fit(
    x = as.matrix(train_data[, c("Glucose", "BMI")]),
    y = as.numeric(train_data$Outcome) - 1,
    epochs = 300,
    batch_size = 100,
    validation_split = 0.2
  )
  
  train_metrics <- model %>% evaluate(
    x = as.matrix(train_data[, c("Glucose", "BMI")]),
    y = as.numeric(train_data$Outcome) - 1
  )
  
  test_metrics <- model %>% evaluate(
    x = as.matrix(test_data[, c("Glucose", "BMI")]),
    y = as.numeric(test_data$Outcome) - 1
  )
  
  return(c(neurons, learning_rate, train_metrics["accuracy"], test_metrics["accuracy"]))
}

# Model parameters
neurons_list <- c(4, 8, 16)
learning_rates <- c(0.001, 0.01)

# Matrix for results
results <- matrix(NA, nrow = length(neurons_list) * length(learning_rates), ncol = 4)
colnames(results) <- c("Neurons", "Learning Rate", "Train Accuracy", "Test Accuracy")

# Running the models and extracting results
index <- 1
for (neurons in neurons_list) {
  for (lr in learning_rates) {
    result <- train_and_evaluate_model(neurons, lr)
    results[index, ] <- result
    index <- index + 1
  }
}

# Summary table of results
summary_table <- as.data.frame(results)
print(summary_table)




